libyul/optimiser/CommonSwitchCasePrefixMover.cpp

```diff
@@ -0,0 +1,161 @@
+
+optional<Statement> tryExtractFirstStatement(Identifier const* _identifier, vector<Case>& _cases)
+{
+	Statement& referenceStatement = _cases.front().body.statements.front();
+
+	if (_identifier)
+	{
+		Assignments assignments;
+		visit(assignments, referenceStatement);
+		if (assignments.names().count(_identifier->name))
+			return nullopt;
+	}
+
+	if (!ranges::all_of(
+		_cases | ranges::cpp20::views::drop(1),
+		[&](Case& _case) {
+			return !_case.body.statements.empty() &&
+				SyntacticallyEqual{}(referenceStatement, _case.body.statements.front());
+		}
+	))
+		return nullopt;
+
+	if (VariableDeclaration const* referenceVarDecl = get_if<VariableDeclaration>(&referenceStatement))
+	{
+		for (Case& switchCase: _cases | ranges::cpp20::views::drop(1))
+		{
+			VariableDeclaration* varDecl = std::get_if<VariableDeclaration>(&switchCase.body.statements.front());
+			yulAssert(varDecl, "");
+			yulAssert(varDecl->variables.size() == referenceVarDecl->variables.size(), "");
+
+			static constexpr auto nameFromTypedName = [](TypedName const& _name) { return _name.name; };
+			IdentifierReplacer replacer{ranges::views::zip(
+				varDecl->variables | ranges::views::transform(nameFromTypedName),
+				referenceVarDecl->variables | ranges::views::transform(nameFromTypedName)
+			) | ranges::to<std::map<YulString, YulString>>};
+			vector<Statement> newBody;
+			for (auto const& stmt: switchCase.body.statements | ranges::cpp20::views::drop(1))
+				newBody.emplace_back(replacer.translate(stmt));
+			switchCase.body.statements = move(newBody);
+		}
+		Statement result = move(referenceStatement);
+		eraseFirst(_cases.front().body.statements);
+		return result;
+	}
+	else
+	{
+		yulAssert(!holds_alternative<FunctionDefinition>(referenceStatement), "FunctionHoister is required.");
+		Statement result = move(referenceStatement);
+		for (Case& switchCase: _cases)
+			eraseFirst(switchCase.body.statements);
+		return result;
+	}
+}
+
+}
+
+void CommonSwitchCasePrefixMover::operator()(Block& _block)
+{
+	util::iterateReplacing(
+		_block.statements,
+		[&](Statement& _s) -> optional<vector<Statement>>
+		{
+			visit(_s);
+			if (Switch* switchStatement = get_if<Switch>(&_s))
+			{
+				yulAssert(!switchStatement->cases.empty(), "");
+
+				Identifier const* identifier = std::get_if<Identifier>(switchStatement->expression.get());
+				if (!identifier && !holds_alternative<Literal>(*switchStatement->expression.get()))
+					return {};
+
+				// We need to be able to tell how the default case behaves.
+				if (switchStatement->cases.back().value)
+					return {};
+
+				vector<Statement> result;
+				while (!switchStatement->cases.front().body.statements.empty())
+					if (auto extractedStatement = tryExtractFirstStatement(identifier, switchStatement->cases))
+						result.emplace_back(move(*extractedStatement));
+					else
+						break;
+
+				if (!result.empty())
+				{
+					result.emplace_back(move(_s));
+					return result;
+				}
+			}
+			return {};
```

Comment:
All these `return {}` statements look a bit as if you were returning an empty vector. I think it would be better to return `nullopt` because it looks less ambiguous:
```suggestion
			return nullopt;
```